### PR TITLE
Support for multiple namespaces with different access modes

### PR DESCRIPTION
**Pull request by NateRoiger**
Signed-off-by: Nate Roiger <nate.roiger@hpe.com>

### CODE REVIEW

File: api/v1alpha1/lustrefilesystem_webhook.go [hunk 1]
```diff
@@ -115,14 +114,28 @@ func (r *LustreFileSystem) validateMountRoot() *field.Error {
...
 	lustrefilesystemlog.Info("validate update", "name", r.Name)
 
-	obj := old.(*LustreFileSystem)
-	// Allow metadata to be updated, for things like finalizers,
-	// ownerReferences, and labels, but do not allow Spec to be updated.
-	if !reflect.DeepEqual(r.Spec, obj.Spec) {
-		return field.Invalid(field.NewPath("spec"), r.Spec, "specification is immutable")
+	old := obj.(*LustreFileSystem)
+
+	// Prevent the immutable fields from being update
```
Comment (bdevcich-hpe) [inline api/v1alpha1/lustrefilesystem_webhook.go:122]:
```suggestion
	// Prevent the immutable fields from being updated
```

File: controllers/lustrefilesystem_controller.go [hunk 2]
```diff
@@ -74,120 +69,209 @@ type LustreFileSystemReconciler struct {
...
 	}
 
-	ownerRef := metav1.NewControllerRef(
-		&fs.ObjectMeta,
-		schema.GroupVersionKind{
-			Group:   v1alpha1.GroupVersion.Group,
-			Version: v1alpha1.GroupVersion.Version,
-			Kind:    "LustreFileSystem",
-		})
+	// Iterative over the access modes in the specification. For each namespace in that mode
```
Comment (bdevcich-hpe) [inline controllers/lustrefilesystem_controller.go:115]:
```suggestion
	// Iterate over the access modes in the specification. For each namespace in that mode,
```

File: controllers/lustrefilesystem_controller_test.go [hunk 3]
```diff
@@ -59,136 +52,185 @@ var _ = Describe("LustreFileSystem Controller", func() {
...
-			time.Sleep(10 * time.Second)
+			Expect(k8sClient.Create(context.TODO(), ns)).Should(Succeed())
 		})
 
-		PIt("Lifecycle of the PVC", func() {
+		BeforeEach(func() {
+			// For some reason envtest never actually deletes a namespace, so instead
+			// managing the namespace for each test case, it is done once in the BeforeAll()
+			// Expect(k8sClient.Create(context.TODO(), ns)).Should(Succeed())
+		})
```
Comment (bdevcich-hpe) [inline controllers/lustrefilesystem_controller_test.go:112]: Delete this and move the comment to the BeforeAll?

File: controllers/lustrefilesystem_controller_test.go [hunk 4]
```diff
@@ -59,136 +52,185 @@ var _ = Describe("LustreFileSystem Controller", func() {
...
+
+					return k8sClient.Update(context.TODO(), fs)
+				}).Should(Succeed())
+
+				Eventually(func(g Gomega) map[string]v1alpha1.LustreFileSystemNamespaceStatus {
+					g.Expect(k8sClient.Get(context.TODO(), client.ObjectKeyFromObject(fs), fs)).Should(Succeed())
+					return fs.Status.Namespaces
+				}).ShouldNot(HaveKey(namespace))
+
+				// envtest is a piece of shit and doesn't support deletion of PV/PVC resources
```
Comment (bdevcich-hpe) [inline controllers/lustrefilesystem_controller_test.go:210]: Maybe reword this 😉

**Review by bdevcich-hpe — APPROVED** [review]

File: api/v1alpha1/lustrefilesystem_webhook.go (hunk 1 above)
Comment (ajfloeder) [inline api/v1alpha1/lustrefilesystem_webhook.go:122]: s/update/updated

**Review by ajfloeder — APPROVED** [review]

File: controllers/lustrefilesystem_controller.go [hunk 5]
```diff
@@ -197,52 +281,55 @@ func createOrUpdatePersistentVolume(ctx context.Context, r *LustreFileSystemReco
...
-	mutateFn := func() error {
-		pvc.ObjectMeta.SetOwnerReferences([]metav1.OwnerReference{*ownerRef})
-
-		pvc.Spec.StorageClassName = &fs.Spec.StorageClassName
-		// Reserve this PVC for the matching PV.
-		pvc.Spec.VolumeName = fs.Name + PersistentVolumeSuffix
-
-		pvc.Spec.AccessModes = []corev1.PersistentVolumeAccessMode{
-			corev1.ReadWriteMany,
+	log.FromContext(ctx).Info("Deleting PersistentVolume", "object", client.ObjectKeyFromObject(pvc).String())
```
Comment (roehrich-hpe) [inline controllers/lustrefilesystem_controller.go:307]: The log message says it's a persistent volume, but you're deleting a PVC.
And in the log below you have these reversed.

**Review by bdevcich-hpe — APPROVED** [review]
